README
AAlib-Ruby brings graphics to the text terminal.  AAlib-Ruby provides both a
graphics context rendered as ascii-art and keyboard and mouse input.
AAlib-Ruby supports a number of text-only display drivers such as Curses,
SLang, and X11.

This is a DL based wrapper around the AA-lib C library.  C struct alignment
issues may cause problems on certain platforms when accessing variables of
some objects.

Author::  Patrick Mahoney (mailto:pat@polycrystal.org)
Copyright:: Copyright (c) 2007 Patrick Mahoney
License:: Distributes under the same terms as Ruby

== Usage

First, AA-lib must be initialized with AAlib.init or AAlib.autoinit,
returning an AAlib::Context.  Two buffers are maintained: the image buffer
and the text buffer.  AAlib::Context#putpixel is the primary method to draw
on the image buffer.  AAlib::Context#render converts the image buffer to
text, writing the result to the text buffer.  Text may be written directly to
the text buffer using AAlib::Context#puts and similar.  Note that text
written this way may be overwritten after a call to AAlib::Context#render.
Finally, AAlib::Context#flush writes the text buffer to the screen.

=== Example

  require 'aalib'

  hp = AAlib::HardwareParams.new
  rp = AAlib::RenderParams.new

  AAlib.parseoptions(hp, rp) or abort AAlib.help
  aa = AAlib.autoinit(hp) or abort "failed to initialize AA-lib"
  begin
    aa.autoinitkbd  # set up keyboard support

    # Fill screen with diagonal gradient
    width = aa.imgwidth
    height = aa.imgheight
    height.times do |y|
      width.times do |x|
        aa.putpixel(x, y, 127*(x.to_f/width) + 127*(y.to_f/height))
      end
    end

    rp.randomval = 25
    aa.render(rp)

    msg = ' AA-lib: the ascii-art library '
    blank = ' ' * msg.size
    attr = AAlib::Attr::BOLD
    col = (aa.scrwidth - msg.size)/2
    row = aa.scrheight/2
    aa.puts(col, row - 1, attr, blank)
    aa.puts(col, row, attr, msg)
    aa.puts(col, row + 1, attr, blank)
    aa.flush

    aa.getkey  # wait for any key to exit
  ensure
    aa.close
  end


ChangeLog
Wed Oct 24 13:09:13 MDT 2007  pat@polycrystal.org
	* Supports most of AA-lib

	* Supports backconvert hack


spec/aalib_spec.rb
require File.join(File.dirname(__FILE__), 'spec_helper')
require 'aalib'

describe "AAlib.autoinit" do
  # We can't check this so much because it might initialize Curses on our terminal
  #
  it "raises TypeError with wrong hardware_params" do
    clo = lambda { AAlib.autoinit(Object.new) }
    clo.should raise_error(TypeError)
  end
end

describe "AAlib.init" do
  before :each do
    @hp = AAlib::HardwareParams.new
  end

  # AAlib.init checks its arguments in violation of duck typing, but incorrect
  # arguments can lead to segmentation faults if AAlib calls the C-library
  # functions with incorrect arguments.

  it "raises TypeError with wrong driver" do
    clo = lambda { AAlib.init(Object.new) }
    clo.should raise_error(TypeError)
  end

  it "raises TypeError with wrong hardware_params" do
    clo = lambda { AAlib.init(AAlib.memory_driver, Object.new) }
    clo.should raise_error(TypeError)
  end

  it "returns an AAlib::Context with memory driver" do
    AAlib.init(AAlib.memory_driver, @hp).should be_an_instance_of(AAlib::Context)
  end

  it "returns an AAlib::Context with memory driver and no hardware_params" do
    AAlib.init(AAlib.memory_driver).should be_an_instance_of(AAlib::Context)
  end

  it "returns an AAlib::Context when called with save driver" do
    opts = AAlib::SaveData.new("test.txt", "text")
    context = AAlib.init(AAlib.save_driver, @hp, opts)
    context.should be_an_instance_of(AAlib::Context)
  end

  it "raises ArgumentError when called with save driver and no driver_opts" do
    clo = lambda { AAlib.init(AAlib.save_driver, @hp, nil) }
    clo.should raise_error(ArgumentError)
  end

  it "raises TypeError when called with save driver and wrong driver_opts" do
    opts = AAlib::SaveFormat.find("text")
    clo = lambda { AAlib.init(AAlib.save_driver, @hp, opts) }
    clo.should raise_error(TypeError)
  end
end

# describe "AAlib.color_from_rgb" do
#   it "converts an RGB color to greyscale" do
#     AAlib.color_from_rgb(0, 0, 0).should == 0
#     AAlib.color_from_rgb(255, 255, 255).should == 255
#     AAlib.color_from_rgb(255, 0, 0).should == 29
#     AAlib.color_from_rgb(0, 255, 0).should == 58
#     AAlib.color_from_rgb(0, 0, 255).should == 10
#     AAlib.color_from_rgb(47, 57, 98).should == 22
#   end
# 
#   it "behaves stupidly when args are larger than 255" do
#     AAlib.color_from_rgb(256, 0, 0).should == 30
#   end
# end

describe "AAlib.help" do
  it "returns a large help string" do
    AAlib.help.should be_an_instance_of(String)
  end
end

describe "AAlib.formats" do
  it "returns an array of formats" do
    AAlib.formats.should be_an_instance_of(Array)
  end

  it "returns an array of AAlib::SaveFormat objects" do
    AAlib.formats.each do |driver|
      driver.should be_an_instance_of(AAlib::SaveFormat)
    end
  end
end

describe "AAlib.drivers" do
  it "returns an array of drivers" do
    AAlib.drivers.should be_an_instance_of(Array)
  end

  it "returns an array of AAlib::Driver objects" do
    AAlib.drivers.each do |driver|
      driver.should be_an_instance_of(AAlib::Driver)
    end
  end
end

describe "AAlib.save_driver" do
  it "returns the driver to use for saving to disk" do
    driver = AAlib.save_driver
    driver.should be_an_instance_of(AAlib::Driver)
    driver.shortname.should == 'save'
    driver.name.should == 'Special driver for saving to files'
  end
end

describe "AAlib.memory_driver" do
  it "returns the driver to use for an in-memory context" do
    driver = AAlib.memory_driver
    driver.should be_an_instance_of(AAlib::Driver)
    driver.shortname.should == 'mem'
    # FIXME: may depend on AA-lib version?
    driver.name.should == 'Dummy memory driver 1.0'
  end
end

describe "AAlib.parseoptions" do
  before :each do
    @hp = AAlib::HardwareParams.new
    @rp = AAlib::RenderParams.new
  end

  it "parses commandline options, setting values in RenderParams" do
    argv = %w{-gamma 2.0 -contrast 5 -bright 100}
    AAlib.parseoptions(@hp, @rp, argv).should be_true
    @rp.gamma.should be_close(2.0, 0.01)
    @rp.contrast.should == 5
    @rp.bright.should == 100
  end

  it "parses commandline options, setting values in HardwareParams" do
    argv = %w{-width 101 -height 57}
    AAlib.parseoptions(@hp, @rp, argv).should be_true
    @hp.width.should == 101
    @hp.height.should == 57
  end

  it "leaves untouched other options" do
    argv = %w{--verbose -gamma 2.0 -contrast 5 -bright 100 -other-option 70}
    AAlib.parseoptions(@hp, @rp, argv).should be_true
    argv.should == %w{--verbose -other-option 70}
  end

  it "uses ARGV by default" do
    ARGV.replace %w{--verbose -gamma 2.0 -contrast 5 -bright 100 -other-option 70}
    AAlib.parseoptions(@hp, @rp).should be_true
    ARGV.should == %w{--verbose -other-option 70}

    @rp.gamma.should be_close(2.0, 0.01)
    @rp.contrast.should == 5
    @rp.bright.should == 100
  end

  it "raises TypeError with wrong hardware_params" do
    clo = lambda { AAlib.parseoptions(Object.new, @rp) }
    clo.should raise_error(TypeError)
  end

  it "raises TypeError with wrong hardware_params" do
    clo = lambda { AAlib.parseoptions(@hp, Object.new) }
    clo.should raise_error(TypeError)
  end
end


spec/spec_helper.rb
$LOAD_PATH.unshift File.join(File.dirname(__FILE__), '..', 'lib')
require 'aalib'

def draw_diagonal_gradient(context)
  width = context.imgwidth
  height = context.imgheight
  height.times do |y|
    width.times do |x|
      context.putpixel(x, y, 127*(x.to_f/width) + 127*(y.to_f/height))
    end
  end
end

def puts_hello_world(context)
  msg = '  AA-lib: the ascii-art library  '
  blank = ' ' * msg.size
  attr = AAlib::Attr::BOLD
  col = (context.scrwidth - msg.size)/2
  row = context.scrheight/2
  context.puts(col, row - 1, attr, blank)
  context.puts(col, row, attr, msg)
  context.puts(col, row + 1, attr, blank)
end


spec/aalib_save_spec.rb
require File.join(File.dirname(__FILE__), 'spec_helper')
require 'aalib'

describe "The save procedure" do
  before :each do
    @file = File.join(File.dirname(__FILE__), "scratch", "test_save")
  end

  after :each do
    # File.delete(@file) if File.exists?(@file)
  end

  it "saves a graphics context as text using the save driver" do
    format = AAlib::SaveFormat.find("text")
    data = AAlib::SaveData.new(@file, format)

    hp = AAlib::HardwareParams.new
    rp = AAlib::RenderParams.new

    context = AAlib.init(AAlib.save_driver, hp, data)
    draw_diagonal_gradient(context)
    context.render(rp)
    puts_hello_world(context)
    context.flush

    context.close

    efile = File.join(File.dirname(__FILE__), "hello_world.txt")
    expected = File.open(efile) { |f| f.read }
    result = File.open(@file) { |f| f.read }

    result.should == expected
  end
end


example/hello_world.rb
require 'aalib'

hp = AAlib::HardwareParams.new
rp = AAlib::RenderParams.new

AAlib.parseoptions(hp, rp) or abort AAlib.help
aa = AAlib.autoinit(hp) or abort "failed to initialize AA-lib"
begin
  aa.autoinitkbd  # set up keyboard support

  # Fill screen with diagonal gradient
  width = aa.imgwidth
  height = aa.imgheight
  height.times do |y|
    width.times do |x|
      aa.putpixel(x, y, 127*(x.to_f/width) + 127*(y.to_f/height))
    end
  end

  rp.randomval = 25
  aa.render(rp)

  msg = ' AA-lib: the ascii-art library '
  blank = ' ' * msg.size
  attr = AAlib::Attr::BOLD
  col = (aa.scrwidth - msg.size)/2
  row = aa.scrheight/2
  aa.puts(col, row - 1, attr, blank)
  aa.puts(col, row, attr, msg)
  aa.puts(col, row + 1, attr, blank)
  aa.flush

  aa.getkey  # wait for any key to exit
ensure
  aa.close
end


example/save.rb
require 'aalib'

hp = AAlib::HardwareParams.new
rp = AAlib::RenderParams.new

AAlib.parseoptions(hp, rp) or abort AAlib.help

file = "aalib-save-demo.txt"
savedata = AAlib::SaveData.new(file, "text", AAlib::SaveFormat::USE_PAGES)
aa = AAlib.init(AAlib.save_driver, hp, savedata)

aa or abort "failed to initialize AA-lib"

begin
  # Fill screen with diagonal gradient
  width = aa.imgwidth
  height = aa.imgheight
  height.times do |y|
    width.times do |x|
      aa.putpixel(x, y, 127*(x.to_f/width) + 127*(y.to_f/height))
    end
  end

  rp.randomval = 25
  aa.render(rp)

  msg = '  AA-lib: the ascii-art library  '
  blank = ' ' * msg.size
  attr = AAlib::Attr::BOLD
  col = (aa.scrwidth - msg.size)/2
  row = aa.scrheight/2
  aa.puts(col, row - 1, attr, blank)
  aa.puts(col, row, attr, msg)
  aa.puts(col, row + 1, attr, blank)

  if File.exists?(file)
    print "File exists: \"#{file}\"  Will not overwrite.\n"
  else
    print "Saving \"#{file}\" ... "
    aa.flush  # Saves file to disk
    print "ok.\n"
  end
ensure
  aa.close
end
